packages/lit-virtualizer/src/lib/repeat.ts

```diff
@@ -0,0 +1,147 @@
+import { directive, NodePart, createMarker, TemplateResult } from 'lit-html';
+import { VirtualRepeater } from './uni-virtualizer/lib/VirtualRepeater';
+import { Layout } from './uni-virtualizer/lib/layouts/Layout';
+
+/**
+ * Mixin for VirtualRepeater and VirtualScroller. This mixin overrides the generic
+ * methods in those classes to provide lit-specific implementations of element
+ * creation and manipulation.
+ * 
+ * This mixin implements child recycling, so children can be reused after removal
+ * from the DOM.
+ */
+export const LitMixin = Superclass => class extends Superclass {
+  // NodeParts that are available for reuse.
+  _pool: Array<NodePart>;
+
+  // Method for generating each item's DOM.
+  _template: (item: any, index?: number) => TemplateResult;
+
+  // Children are rendered into this part.
+  // The host of the directive that constructs the scroller.
+  _hostPart: NodePart;
+  
+  /**
+   * TODO @straversi: This is not a true mixin. Mixins do have information
+   * about potential superclasses. This class appears to know the constructor
+   * arguments of an unknown superclass. It also overrides methods on an
+   * unknown superclass.
+   */
```

Comment:
The current factoring employs a mixin pattern that @justinfagnani championed and that we've used extensively in our JS code, since JS doesn't have first-class mixin support. If TS offers a better pattern, then we should certainly use it.

I will say, though, that the aim of this factoring wasn't to extend an arbitrary superclass. Rather, it was to address a layering problem. The generic `uni-virtual` implementation has two base classes (`VirtualRepeater` and `VirtualScroller`); we wanted lit-specific subclasses of each. Since the lit-specific customizations in each case were identical, factoring them into a single mixin made sense. That mixin is designed to extend not just any class, but only `VirtualReapeater` or a subclass thereof (like `VirtualScroller`).

Reply:
The TS pattern is mostly about how the types are annotated and/or inferred. The only code change necessary is that constructors have to take arbitrary arguments, ie `constructor(...args)`